contracts/ExampleTrustcoin2.sol

```diff
@@ -64,87 +58,68 @@ contract ExampleTrustcoin2 is ERC20TokenInterface, SafeMath {
     return allowed[_owner][_spender];
   }
 
   //
   //  Migration methods
   //
 
-  /**
-   *  Changes the owner for the migration behaviour
-   *  @param _master Address of the migration controller
-   */
-  function changeMigrationMaster(address _master) onlyFromMigrationMaster external {
+  // See MigratedTokenInterface
+  function migrateOldTokens(address _from, uint256 _value) external {
+    if (!allowIncomingMigrations) throw;
+    if (_value == 0) throw;
+    totalSupply = safeAdd(totalSupply, _value);
+    balances[_from] = safeAdd(balances[_from], _value);
+    IncomingMigration(_from, _value);
+  }
+
+  // See MigratableTokenInterface
+  function changeMigrationMaster(address _master) external {
     if (_master == 0) throw;
     migrationMaster = _master;
   }
 
-  /**
-   *  Sets the address of the new token contract, so we know who to
-   *  accept discardTokens() calls from, and enables token migrations
-   *  @param _newTokenAddress Address of the new Trustcoin contract
-   */
-  function setNewTokenAddress(address _newTokenAddress) onlyFromMigrationMaster external {
+  // See MigratableTokenInterface
+  function setNewTokenAddress(address _newTokenAddress) external {
     if (newTokenAddress != 0) throw; // Ensure we haven't already set the new token
     if (_newTokenAddress == 0) throw;
     newTokenAddress = _newTokenAddress;
+    allowOutgoingMigrationsUntil = (now + 26 weeks); // Only allow migrations for the next six months
   }
 
-  /**
-   *  Migrates the specified token balance from msg.sender in the old contract
-   *  to the new contract
-   *  @param _value Number of tokens to be migrated
-   */
-  function migrateOldTokens(uint256 _value) external {
-    if (!allowIncomingMigrations) throw;
+  // See MigratableTokenInterface
+  function discardTokens(uint256 _value) external {
+    if (newTokenAddress == 0) throw; // Ensure that we have set the new token
+    if (now < allowOutgoingMigrationsUntil) throw;
     if (_value == 0) throw;
-    Trustcoin(oldToken).discardTokens(msg.sender, _value);
-    totalSupply = safeAdd(totalSupply, _value);
-    balances[msg.sender] = safeAdd(balances[msg.sender], _value);
-    IncomingMigration(msg.sender, _value);
+    if (_value > balances[msg.sender]) throw;
+    balances[msg.sender] = safeSub(balances[msg.sender], _value);
+    totalSupply = safeSub(totalSupply, _value);
+    totalMigrated = safeAdd(totalMigrated, _value);
+    MigratedTokenInterface(newTokenAddress).migrateOldTokens(msg.sender, _value);
+    OutgoingMigration(msg.sender, _value);
   }
 
-  /**
-   *  Ends the possibility for any more tokens to be migrated from the old contract
-   *  to the new one
-   */
-  function finalizeMigration() onlyFromMigrationMaster external {
+  // Ends the possibility for any more tokens to be migrated from the old contract
+  // to the new one
+  function finalizeMigration() onlyFromMigrationMaster(msg.sender) external {
```

Comment:
consider renaming to `finalizeIncomingMigration` (I got confused.. :nerd_face: )